### DISCUSSION TITLE
Rename 'By status' to 'By urgency'

### DISCUSSION THREAD

**Discussion by puelp** [iSoron/uhabits#921] [Feature Requests]
'By status' is confusing because it could mean anything. I had to tap on it and play around with it to understand that it meant sorting the most overdue tasks to the top/bottom.

I feel that 'By urgency' would be a more apt label.

So basically...:

Sort >
Manually
By name
By color
By score
By urgency

**Comment by iSoron** [comment]
Thank you for the feature request, @puelp. I agree that "by status" is not entirely clear, but I personally don't think "by urgency" is a much better description. If anyone else disagrees or has a better suggestion, please feel free to leave a comment.

**Reply by Silviopina** [reply]
By priority

**Reply by hiqua** [reply]
That's what I called it initially in https://github.com/iSoron/uhabits/pull/660 so you already know @iSoron's opinion about this option (https://github.com/iSoron/uhabits/pull/660#issuecomment-729309131)

**Comment by hiqua** [comment]
I don't think any other short name will fix this. I'd prefer a toast explaining how the feature works, if we can come up with something short enough.

I think it'd also be fine to explain this in some external documentation, like the FAQ.

I view the sorts as meant to cater to users explicitly looking for them ("I want to sort by color, oh nice it's implemented"), not as features you want to expose as selling-points ("we have sort-by-status, this app is awesome you should install it"). As such, the current situation is fine as it is. And you did end up finding out how it works in the end.

Sorts are also a power-user feature in a way, since they're relatively useless until you have enough habits to fill one screen (12 on my phone). At least I think I'd sort manually if I had such a low amount of habits. I think it's fine for power-user features to be slightly less easy to understand (and again, you did figure it out on your own).

**Comment by drawtheliner** [comment]
**By priority** makes sense for me.

However I just did a test on a fresh install, with nothing checked yet, and it doesn't seem to be working:

Here's the order I get:
1. Daily habit
2. 5 times a week
3. Every 99 days
4. Every 20 days

Shouldn't 3 and 4 be swapped?
I didn't do further tests.

**Reply by hiqua** [reply]
They all need to be checked in this case so the sort doesn't change the positions I'd say.

**Reply by drawtheliner** [reply]
Ah ok. I thought there was some sort of algorithm that helped you tackle habits in the most efficient order possible.

Between a habit I want to do every 20 days and another I want to do every 99 days, the former is clearly priority (since I have more leeway for the habit with longer cycle) . Even if I've done the 20 days one recently, it might still be more efficient to tackle it first 

This kind of information would be very useful to help people keep with their habits, imho.

Should I make a separate feature request for this type of priority logic,?

**Reply by hiqua** [reply]
> the former is clearly priority

Not sure I follow the reasoning, but if there's currently no data, I'd consider the priority undefined or at least arbitrary depending on the user. Just because a habit is supposed to be done more frequently, doesn't mean it's more important.

> Even if I've done the 20 days one recently, it might still be more efficient to tackle it first

You can combine sorts, e.g. sort both by status and score, which should help for this kind of stuff. IMHO as a user you're better off not pondering too much over which habit to complete though.

> Should I make a separate feature request for this type of priority logic,?

Feel free, there might be existing ones already though. We could also modify the current status sort. You can also have a look at the code directly, it should be relatively simple to change, e.g. start from there: https://github.com/iSoron/uhabits/blob/dev/uhabits-core/src/jvmMain/java/org/isoron/uhabits/core/models/memory/MemoryHabitList.kt#L120

In any case try to specify further what behavior you'd find interesting.

**Reply by drawtheliner** [reply]
> Just because a habit is supposed to be done more frequently, doesn't mean it's more important.

Not more important, just more urgent (in terms of keeping up with your habits).

Maybe an example with shorter time span would help. Let's say you have a Weekly habit and a Monthly habit. You complete both today. Then 6 days from now you are very busy and are only able to do 1 of them. Which do you do?

**Reply by hiqua** [reply]
So you want to sort by frequency?

**Reply by drawtheliner** [reply]
That example was to explain what I mean by priority.

Usually the situation is more complex than that. We have multiple habits, each with different frequencies and with different "last completed" dates. 

For example, a monthly habit will be more urgent (and thus would have higher priority) than a weekly one if it has been neglected long enough. Does that make sense?

There would need to be an algorithm that takes all factors into account.

**Reply by hiqua** [reply]
> There would need to be an algorithm that takes all factors into account.

So sorting by score? If that's still not it you'll have to write it down or implement it I'm afraid :slightly_smiling_face: 

One problem we'll encounter is that priorities might be different depending on the users, I'm not sure there's a heuristic that can apply to, say, 50%+ of users.

**Reply by drawtheliner** [reply]
> So sorting by score?

Maybe! I couldn't find an explanation of what Score is exactly... Do you know where I can find this?

Edit: By the way, I found a flaw in the (very superficial) [test I did](https://github.com/iSoron/uhabits/discussions/921#discussioncomment-2769473) earlier. I did it on a fresh install, with no checkmarks. This means Loop has no data in order to assess urgency. So maybe Status is what I'm looking for. Or perhaps it's a formula involving that **and** Score (assuming Score relates to how well you've been keeping to a habit frequency goal - just a guess).

**Reply by hiqua** [reply]
Read "How is the score calculated?" on https://github.com/iSoron/uhabits/discussions/689

**Reply by drawtheliner** [reply]
Thank you!

Yeah, so I guess the formula for my proposed metric called "Priority" would involve both, the Status and Score (if I manage to think up a suggestion for an algorithm I'll share it here - I'm not a coder though, so it will be pseudo-code).

> One problem we'll encounter is that priorities might be different depending on the users, I'm not sure there's a heuristic that can apply to, say, 50%+ of users.

Yes, I agree. But I don't see this as a big problem, as this would be optional. The Score metric is also arbitrary - but it's still helpful for most people, even to those that might disagree with it. Also, you can just ignore it (I do).

**Comment by visika** [comment]
I discovered that sorting is a bit more nuanced than I initially thought.
Applying two consecutive sort criteria yields different results for different criteria.
I'll make an example.
Sorting by status puts the unmarked habits at the top and the marked habits at the bottom.
But how are the marked habits at the top ordered?
They are ordered according to the previous selected sorting.
So, applying alphabetic sorting first and status sorting second gives a simple staggered view.
If we want a view that communicates the sense of urgency, or priority if you like, it should be applied sorting by score in the increasing direction (arrow pointing up, apply twice to switch) and sorting by status (arrow pointing down) in succession.
This separates the still undone habits at the top and puts the more neglected at the top of the top.
